packages/runtime-common/realm.ts

```diff
@@ -526,6 +551,223 @@ export class Realm {
+    for (let { href } of operations) {
+      promises.push(this.#adapter.exists(href));
+    }
+    let booleanFlags = await Promise.all(promises);
+    return operations
+      .filter((_, i) => booleanFlags[i])
+      .map(({ href }) => {
+        return {
+          title: 'Resource already exists',
+          detail: `Resource ${href} already exists`,
+          status: 409,
+        };
+      });
+  }
+
+  private lowestStatusCode(errors: ErrorDetails[]): number {
+    let statuses = errors
+      .filter((e) => e.status)
+      .map((e) => e.status) as number[];
+    return Math.min(...statuses);
+  }
+
+  private async handleAtomicOperations(
+    request: Request,
+    requestContext: RequestContext,
+  ): Promise<Response> {
+    let body = await request.text();
+    let json;
+    try {
+      json = JSON.parse(body);
+    } catch (e) {
+      return createResponse({
+        body: JSON.stringify({
+          errors: [
+            {
+              title: 'Invalid atomic:operations format',
+              detail: `Request body is not valid JSON`,
+            },
+          ],
+        }),
+        init: {
+          status: 400,
+          headers: {
+            'content-type': SupportedMimeType.CardJson,
+          },
+        },
+        requestContext,
+      });
+    }
+    let validationErrors = this.validate(json);
+    if (validationErrors.length > 0) {
+      return createResponse({
+        body: JSON.stringify({ errors: validationErrors }),
+        init: { status: 400, headers: { 'content-type': 'application/json' } }, //consolidate to 400
+        requestContext,
+      });
+    }
+    let operations = json['atomic:operations'] as AtomicOperation[];
+
+    let atomicCheckErrors = await this.checkBeforeAtomicWrite(operations);
+    if (atomicCheckErrors.length > 0) {
+      return createResponse({
+        body: JSON.stringify({ errors: atomicCheckErrors }),
+        init: {
+          status: this.lowestStatusCode(atomicCheckErrors),
+          headers: { 'content-type': 'application/json' },
+        },
+        requestContext,
+      });
+    }
+    let sourceFiles = new Map<LocalPath, string>();
+    let instanceFiles = new Map<LocalPath, string>();
+    let sourceOperations = filterOperationsWithSourceData(operations);
+    let instanceOperations = filterOperationsWithCardData(operations);
+    let instanceWriteResults: WriteResult[] = [];
+    let sourceWriteResults: WriteResult[] = [];
+
+    // we intentionally write source files first so that fileSerialization can work
+
+    for (let { data: resource, href } of sourceOperations) {
+      let fileURL = this.paths.fileURL(href);
+      let localPath = this.paths.local(fileURL);
+      let content = resource.attributes?.content as string;
+      sourceFiles.set(localPath, content);
+    }
+
+    if (sourceFiles.size > 0) {
+      sourceWriteResults = await this.write(
+        sourceFiles,
+        request.headers.get('X-Boxel-Client-Request-Id'),
+      );
+    }
```

Comment:
Does it mean that we have to wait for an indexing run after the source files are written before we write the card data files?

Reply:
Yes. This is this behavior unfortunately. A failure to write one of the modules will stop the writes of ALL the instances. And modules will remain written as there is no cleanup currently. 

Do you have a proposed way to resolve this? Essentially, `fileSerialization` is the method that creates this dependence between instance and their modules. 

**Side Note**

Sanity-check. I think we already wait for indexing of this already bcos the `update` method in realm-index-updated waits for the `job.done`. Otherwise, we'd have an issue of indexing of the module being incomplete before trying to write the instance.